unittests/FEXLinuxTests/tests/signal/eflags_signal.cpp

```diff
@@ -0,0 +1,303 @@
+  DO_ASM(
+    // Clear OF
+    "mov eax, 0;"
+    "inc eax;"
+  ,
+  // OF should still be cleared.
+  "jno 1f;"
+  "int3;"
+  "1:"
+  );
+}
+
+static void SetOFAndWait(std::atomic<uint32_t> *Futex, std::atomic<uint32_t> *ReadyNotify) {
+  DO_ASM(
+    // Set OF
+    "mov eax, 0x7fffffff;"
+    "inc eax;"
+  ,
+  // OF should still be set.
+  "jo 1f;"
+  "int3;"
+  "1:"
+  );
+}
+
+struct CapturingData {
+  int Signal;
+  uint64_t eflags;
+};
+
+std::optional<CapturingData> from_handler;
+constexpr uint32_t EFL_CF = 0;
+constexpr uint32_t EFL_PF = 2;
+constexpr uint32_t EFL_ZF = 6;
+constexpr uint32_t EFL_SF = 7;
+constexpr uint32_t EFL_OF = 11;
+
+static void CapturingHandler(int signal, siginfo_t *siginfo, void* context) {
+  ucontext_t* _context = (ucontext_t*)context;
+
+  from_handler = {
+    .Signal = signal,
+    .eflags = static_cast<uint64_t>(_context->uc_mcontext.gregs[REG_EFL]),
+  };
+}
+
+using TestHandler = std::function<void(std::atomic<uint32_t> *Futex, std::atomic<uint32_t> *ReadyNotify)>;
+
+static void ThreadHandler(std::atomic<uint32_t> *Mutex, std::atomic<uint32_t> *ReadyNotify, std::atomic<uint32_t> *ThreadID, TestHandler Test) {
+  // Unblock SIGTERM.
+  sigset_t BlockMask{};
+  sigemptyset(&BlockMask);
+  sigaddset(&BlockMask, SIGTERM);
+  sigprocmask(SIG_UNBLOCK, &BlockMask, nullptr);
+
+  // Set up a signal handler for SIGTERM
+  struct sigaction act{};
+  act.sa_sigaction = CapturingHandler;
+  act.sa_flags = SA_SIGINFO;
+  sigaction(SIGTERM, &act, nullptr);
+
+  *ThreadID = ::gettid();
+  Test(Mutex, ReadyNotify);
+}
+
+void WaitForThreadAsleep(uint32_t tid) {
+  std::string Path = "/proc/" + std::to_string(::getpid()) + "/task/" + std::to_string(tid) + "/status";
+  std::ifstream fs{Path, std::fstream::binary};
+  std::string Line;
+
+  while (true) {
+    fs.clear();
+    fs.seekg(0);
+    while (std::getline(fs, Line)) {
+      if (fs.eof()) break;
+
+      if (Line.find("State") == Line.npos) {
+        continue;
+      }
+
+      char State{};
+      if (sscanf(Line.c_str(), "State: %c", &State) == 1) {
+        if (State == 'S') {
+          return;
+        }
+        break;
+      }
+    }
+  }
+}
+void RunTest(uint32_t FlagLocation, uint32_t FlagValue, TestHandler Test) {
+  // Block SIGTERM.
+  sigset_t BlockMask{};
+  sigemptyset(&BlockMask);
+  sigaddset(&BlockMask, SIGTERM);
+  sigprocmask(SIG_BLOCK, &BlockMask, nullptr);
+  std::atomic<uint32_t> Mutex{};
+  std::atomic<uint32_t> ReadyNotify{};
+  std::atomic<uint32_t> ThreadID{};
+
+  std::thread t(ThreadHandler, &Mutex, &ReadyNotify, &ThreadID, Test);
+
+  while (ReadyNotify.load() == 0);
+  // Wait for thread to get in to the futex.
+  WaitForThreadAsleep(ThreadID.load());
+
+  tgkill(::getpid(), ThreadID.load(), SIGTERM);
+
+  t.join();
+
+  REQUIRE(from_handler.has_value());
+  REQUIRE(from_handler.value().Signal == SIGTERM);
+  REQUIRE(((from_handler.value().eflags >> FlagLocation) & 1) == FlagValue);
```

Comment:
REQUIRE will abort any later checks and hence should only be used if those checks can't run properly (e.g. because they'd dereference a nullptr). Using CHECK in other cases will print more information to help diagnose test failures.

```suggestion
  REQUIRE(from_handler.has_value());
  CHECK(from_handler.value().Signal == SIGTERM);
  CHECK(((from_handler.value().eflags >> FlagLocation) & 1) == FlagValue);
```

Reply:
Resolved.